blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala

```diff
@@ -22,147 +22,141 @@ import cats.effect.implicits._
+    responseHeaderTimeout: Duration,
+    requestTimeout: Duration,
+    scheduler: TickWheelExecutor,
+    ec: ExecutionContext,
+    retries: Int,
+    dispatcher: Dispatcher[F],
+)(implicit F: Async[F])
+    extends DefaultClient[F] {
+
+  private[this] val logger = getLogger
+
+  override def run(req: Request[F]): Resource[F, Response[F]] =
+    Hotswap.create[F, Either[Throwable, Response[F]]].flatMap { hotswap =>
+      Resource.eval(retryLoop(req, retries, hotswap))
+    }
+
+  // A light implementation of the Retry middleware.  That needs a
+  // Timer, which we don't have.
+  private def retryLoop(
+      req: Request[F],
+      remaining: Int,
+      hotswap: Hotswap[F, Either[Throwable, Response[F]]],
+  ): F[Response[F]] =
+    hotswap.clear *> // Release the prior connection before allocating the next, or we can deadlock the pool
+      hotswap.swap(respond(req).attempt).flatMap {
+        case Right(response) =>
+          F.pure(response)
+        case Left(_: SocketException) if remaining > 0 && req.isIdempotent =>
+          val key = RequestKey.fromRequest(req)
+          logger.debug(
+            s"Encountered a SocketException on ${key}.  Retrying up to ${remaining} more times."
+          )
+          retryLoop(req, remaining - 1, hotswap)
+        case Left(e) =>
+          F.raiseError(e)
       }
+
+  private def respond(req: Request[F]): Resource[F, Response[F]] = {
+    val key = RequestKey.fromRequest(req)
+    for {
+      requestTimeoutF <- scheduleRequestTimeout(key)
+      preparedConnection <- prepareConnection(key)
+      (conn, responseHeaderTimeoutF) = preparedConnection
+      timeout = responseHeaderTimeoutF.race(requestTimeoutF).map(_.merge)
+      responseResource <- Resource.eval(runRequest(conn, req, timeout))
+      response <- responseResource
+    } yield response
+  }
+
+  private def prepareConnection(key: RequestKey): Resource[F, (A, F[TimeoutException])] = for {
+    conn <- borrowConnection(key)
+    responseHeaderTimeoutF <- addResponseHeaderTimeout(conn)
+  } yield (conn, responseHeaderTimeoutF)
+
+  private def borrowConnection(key: RequestKey): Resource[F, A] =
+    Resource.makeCase(manager.borrow(key).map(_.connection)) {
+      case (conn, ExitCase.Canceled) =>
+        // Currently we can't just release in case of cancellation, because cancellation clears the Write state of Http1Connection, so it might result in isRecycle=true even if there's a half-written request.
+        manager.invalidate(conn)
+      case (conn, _) => manager.release(conn)
     }
 
-  @deprecated("Preserved for binary compatibility", "0.23.8")
-  private[blaze] def makeClient[F[_], A <: BlazeConnection[F]](
-      manager: ConnectionManager[F, A],
-      responseHeaderTimeout: Duration,
-      requestTimeout: Duration,
-      scheduler: TickWheelExecutor,
-      ec: ExecutionContext,
-  )(implicit F: Async[F]): Client[F] =
-    makeClient(manager, responseHeaderTimeout, requestTimeout, scheduler, ec, 0)
```

Comment:
We need a Dispatcher, so I can't drag this along anymore.  It's private, so, whatever.